src/net/server/bytearrstream.rs

```diff
@@ -0,0 +1,451 @@
+            // visit `size_in_bytes` buffers in order to fill `out_bytes`
+            loop {
+                let remaining_bytes = out_bytes.capacity() - out_bytes.len();
+                if remaining_bytes == 0 {
+                    break;
+                }
+                // current `buffer` has more bytes than needed
+                if self.buffered[0].len() > remaining_bytes {
+                    let remaining = self.buffered[0].split_to(remaining_bytes);
+                    out_bytes.extend_from_slice(&remaining);
+                } else {
+                    // current `buffer` has <= bytes needed, remove and append to out_bytes
+                    out_bytes.extend_from_slice(&self.buffered.pop_front().unwrap());
+                }
+            }
+            // reduce size of total buffers accordingly
+            self.buffered_size -= self.size_in_bytes;
+            Some(Bytes::from(out_bytes))
+        }
+    }
+
+    fn push_back(&mut self, buf: Bytes) {
+        self.buffered_size += u32::try_from(buf.len()).unwrap();
+        self.buffered.push_back(buf);
+    }
+}
+
+impl Stream for ByteArrStream {
+    type Item = Result<Bytes, std::io::Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            // if we have enough bytes buffered, return them
+            if let Some(bytes) = self.pop_front() {
+                return Poll::Ready(Some(Ok(bytes)));
+            }
+            // if we need more bytes, poll the body
+            let mut this = self.as_mut().project();
+            match ready!(this.body.as_mut().poll_next(cx)) {
+                // if body is expended, but we have some bytes leftover, error
+                None if *this.buffered_size > 0 => {
+                    return Poll::Ready(Some(Err(std::io::Error::new::<BoxError>(
+                        std::io::ErrorKind::WriteZero,
+                        format!(
+                            "expected body to align on size {}, but has insufficient bytes {}",
+                            *this.size_in_bytes, *this.buffered_size
+                        )
+                        .into(),
+                    ))));
+                }
+                // if body is finished, this stream is finished
+                None => return Poll::Ready(None),
+                // if body produces error, forward the error
+                Some(Err(err)) => {
+                    return Poll::Ready(Some(Err(std::io::Error::new::<BoxError>(
+                        std::io::ErrorKind::UnexpectedEof,
+                        err.into(),
+                    ))));
+                }
+                // if body has more bytes, push it into the buffer and loop
+                Some(Ok(bytes)) => {
+                    self.push_back(bytes);
+                }
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl<B: HttpBody<Data = Bytes, Error = hyper::Error> + Send + 'static> FromRequest<B>
```

Comment:
> Bytes is a 32 byte pointer and it would be a waste of space to use it to pass around 4 or 8 byte chunks

this makes sense, but the reason it's done this way right now is to try to avoid copies. Are you saying that we _should_ be copying out of upstream `Bytes` array? and so this would actually be a `Stream<Item = Vec<u8>>`? we can't use arrays because len is not known at compile time.

> Right now ByteArrStream does not support chuhks smaller that minimum

not sure I quite understand. What do you mean by "chunks smaller than minimum"?

Reply:
Would I be able to ask `ByteArrStream` to give me chunks of size 400 bytes if the whole input is 100 bytes?

> Are you saying that we should be copying out of upstream Bytes array? and so this would actually be a Stream<Item = Vec<u8>>?

no, all I am saying is slicing 1Gb input into chunks of 4 bytes is too wasteful. When chunk size on the heap is smaller than the pointer size, it looks very inefficient to me. We would need to be able to carve out chunks of much much larger size than 4 bytes as long as they are aligned with the field value size.

Reply:
gotcha, i see what you're saying. and your other point makes sense now as well. I can add a TODO to figure that out

Reply:
added: https://github.com/private-attribution/ipa/pull/344/files#diff-de21be8a069c815dcad06256ba2473fcb77d6081a62e09f29e367913e38029f3R38